### PR TITLE
Adding certstream (https://github.com/CaliDog/certstream-python)

### PR DESCRIPTION
**Pull request by cgi1**
Adding Certificate Transparency Log (CTL) Collector/Parser

### CODE REVIEW

File: intelmq/bots/parsers/certstream/parser_certstream.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                    self.send_message(event)
+
+        if 'chain' in data:
+            for chain_element in data['chain']:
+                pass # ToDo: Add chains
+
+        self.acknowledge_message()
+
+
+if __name__ == "__main__":
```
Comment (wagner-certat) [inline intelmq/bots/parsers/certstream/parser_certstream.py]: same as above

File: intelmq/bots/parsers/certstream/parser_certstream.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    def process(self):
+        report = self.receive_message()
+
+        raw_report = utils.base64_decode(report.get('raw'))
+        try:
+
+            json_report = json.loads(raw_report)
+        except:
+            self.logger.exception("Failed to convert raw_report to json!")
```
[the remark this answers is not in the document]
Comment (cgi1) [reply intelmq/bots/parsers/certstream/parser_certstream.py]: Did not had the problem yet - But if so the stacktrace would be nice.

**Comment by cgi1** [comment]
@wagner-certat:  Pls check it again. 

Can not reproduce `Error connecting to CertStream - cafile, capath and cadata cannot be all omitted`

Other suggestions are fixed.

**Comment by cgi1** [comment]
Reopend

File: intelmq/bots/collectors/certstream/collector_certstream.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+from intelmq.lib.bot import Bot
+from intelmq.lib.message import Report
+
+
+class CertstreamCollectorBot(Bot):
+
+    def init(self):
+
+        self.logger.debug("Nothing to init.")
```
[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/bots/collectors/certstream/collector_certstream.py]: > 43067ae Removed init message (was a good idea to see that the bot initialized, but anyway)

There are a couple of other log messages at debug level, e.g. `Bot is starting`. What were you able to detect with this additional log message?

File: intelmq/bots/parsers/certstream/parser_certstream.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/bots/parsers/certstream/parser_certstream.py]: Is the reason/position shown in the stacktrace?

[the remark this answers is not in the document]
Comment (cgi1) [reply intelmq/bots/parsers/certstream/parser_certstream.py]: As far as I remember yes

[the remark this answers is not in the document]
Comment (SYNchroACK) [reply intelmq/bots/parsers/certstream/parser_certstream.py]: @cgi1 replace the following code:
```
try:
    json_report = json.loads(raw_report)
except:
    self.logger.exception("Failed to convert raw_report to json!")
    return
```

with only this line:
```
json_report = json.loads(raw_report)
```

File: intelmq/bots/parsers/certstream/parser_certstream.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            event.add('time.source', datetime.datetime.fromtimestamp(int(data['seen'])).strftime('%Y-%m-%d %H:%M:%S UTC'))
+
+        # ToDo: Check if leaf_cert --> extensions --> subjectAltName is identical to all_domains
+        # ToDo: Check if leaf_cert --> extensions --> extendedKeyUsage is always for Web Server Authentication (if not filter)
+
+        if 'leaf_cert' in data:
+            if 'all_domains' in data['leaf_cert']:
+                for domain in data['leaf_cert']['all_domains']:
+
+                    try:
```
Comment (SYNchroACK) [inline intelmq/bots/parsers/certstream/parser_certstream.py]: From line 46 to 56, I would suggest to replace with the following lines:
```
for value in data['leaf_cert']['all_domains']:
    for key in ['source.fqdn', 'source.ip']:

        try:
            event.add(key, value)
            self.send_message(event)
            break

        except exceptions.InvalidValue:
            continue

    else:
        self.logger.debug("Invalid value (%s) found in 'leaf_cert:all_domains' field. Not a valid ip or domain." % value)
```

**Comment by wagner-certat** [comment]
> Can not reproduce `Error connecting to CertStream - cafile, capath and cadata cannot be all omitted`

I found this related bug: websocket-client/websocket-client#227 See also the linked reports in related software.
But as I have version 0.37.0 here - the bug was fixed in version 0.35.0 - we need to specify the ca path "manually". Similar to what we have for xmpp, but hopefully more generic.

**Comment by wagner-certat** [comment]
I opened the bug CaliDog/certstream-python#5 and the PR CaliDog/certstream-python#6 to fix the issue

**Comment by wagner-certat** [comment]
> We are using the code productive for a few weeks now. What is essentially missing for an add to the repo?

See my last feedback in https://github.com/certtools/intelmq/pull/1120#pullrequestreview-78222596

> I am still missing documentation in docs/Bots.md / docs/Feeds.md

Documentation _is_ essential. `docs/Feeds.md` has been rewritten in the meantime to use `intelmq/etc/feeds.yaml`, but I can convert that myself too.

**Pull request merged**